### PR TITLE
Implement H2O model instance

### LINKED ISSUE #129: Create Model and ModelInstance for H2O

### PR DESCRIPTION
**Pull request by kantoniak**
Implements `H2OModelInstance`, missing parts of `H2OModel` and adds a mock of `H2OModelPayload`.
Closes #129.

### CODE REVIEW

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModel.scala [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
 
-class H2OModel extends Model {
-    //TODO placeholder model - we should create this model based on some config
-    private val genModel: GenModel = MojoModel.load("")
-    //TODO similarly to evaluator in PMMLModel we can use this wrapper to extract signature and score the model in ModelInstance
+class H2OModel(payload: H2OModelPayload) extends Model {
+    // TODO: Should caching strategy be configurable?
+    private val modelReaderBackend = MojoReaderBackendFactory.createReaderBackend(payload.inputStream, CachingStrategy.MEMORY)
+    private val genModel: GenModel = ModelMojoReader.readFrom(modelReaderBackend)
     val modelWrapper: EasyPredictModelWrapper = new EasyPredictModelWrapper(genModel)
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModel.scala:13]: why it is not private?

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala [hunk 2]
```diff
@@ -1,3 +1,73 @@
...
+    } catch {
+      case ex: PredictException =>
+        logger.warn("Got PredictException:", ex)
+        Left(new ModelRunException(ex.toString))
+    }
+  }
+
+  def evaluateRow(row: Map[String, AnyRef]): AbstractPrediction = {
+    logger.info("Evaluate row {}", row)
+    var rowData = new RowData()
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala]: val

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala [hunk 3]
```diff
@@ -1,3 +1,73 @@
...
+  def getTransformer(modelCategory: ModelCategory): (AbstractPrediction => _) = modelCategory match {
+    case ModelCategory.AnomalyDetection => (p: AbstractPrediction) => p.asInstanceOf[AnomalyDetectionPrediction].isAnomaly
+    case ModelCategory.Binomial         => (p: AbstractPrediction) => p.asInstanceOf[BinomialModelPrediction].labelIndex
+    case ModelCategory.Multinomial      => (p: AbstractPrediction) => p.asInstanceOf[MultinomialModelPrediction].labelIndex
+    case ModelCategory.Ordinal          => (p: AbstractPrediction) => p.asInstanceOf[OrdinalModelPrediction].labelIndex
+    case ModelCategory.Clustering       => (p: AbstractPrediction) => p.asInstanceOf[ClusteringModelPrediction].cluster
+    case ModelCategory.KLime            => (p: AbstractPrediction) => p.asInstanceOf[KLimeModelPrediction].cluster
+    case ModelCategory.CoxPH            => (p: AbstractPrediction) => p.asInstanceOf[CoxPHModelPrediction].value
+    case ModelCategory.Regression       => (p: AbstractPrediction) => p.asInstanceOf[RegressionModelPrediction].value
+    case _ => throw new ModelRunException("ModelCategory " + modelCategory.toString() + " not supported.")
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala]: prefer string interpolation

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala [hunk 4]
```diff
@@ -1,3 +1,73 @@
...
+    logger.info("Evaluate row {}", row)
+    var rowData = new RowData()
+    rowData.putAll(row.asJava)
+    logger.info("Args for H2O evaluator: {}", rowData)
+    val result = modelWrapper.predict(rowData)
+    logger.info("Evaluation result: {}", result)
+    result
+  }
+
+  def getTransformer(modelCategory: ModelCategory): (AbstractPrediction => _) = modelCategory match {
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala]: private

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala [hunk 5]
```diff
@@ -1,3 +1,73 @@
...
+    case ModelCategory.Multinomial      => (p: AbstractPrediction) => p.asInstanceOf[MultinomialModelPrediction].labelIndex
+    case ModelCategory.Ordinal          => (p: AbstractPrediction) => p.asInstanceOf[OrdinalModelPrediction].labelIndex
+    case ModelCategory.Clustering       => (p: AbstractPrediction) => p.asInstanceOf[ClusteringModelPrediction].cluster
+    case ModelCategory.KLime            => (p: AbstractPrediction) => p.asInstanceOf[KLimeModelPrediction].cluster
+    case ModelCategory.CoxPH            => (p: AbstractPrediction) => p.asInstanceOf[CoxPHModelPrediction].value
+    case ModelCategory.Regression       => (p: AbstractPrediction) => p.asInstanceOf[RegressionModelPrediction].value
+    case _ => throw new ModelRunException("ModelCategory " + modelCategory.toString() + " not supported.")
+  }
+
+  def collectOutputs(rows: IndexedSeq[AbstractPrediction]): Map[String, _] = {
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala]: private

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala [hunk 6]
```diff
@@ -1,3 +1,73 @@
...
+    rows.take(1).size match {
+      case 0 => Map[String, Any]()
+      case 1 =>
+        val transformer = getTransformer(modelWrapper.m.getModelCategory)
+        var results = new ListBuffer[Any]()
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala]: val

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala [hunk 7]
```diff
@@ -1,3 +1,73 @@
...
+        rows.foreach(r => results += transformer(r))
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala]: and then here use rows.map(transformer) or something similiar

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/repository/H2OModelPayload.scala [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package pl.touk.nussknacker.prinz.h2o.repository
+
+import java.io.InputStream
+
+// Tip: MOJO readers also support java.net.URL in place of InputStream
+final case class H2OModelPayload(inputStream: InputStream, name: String, version: String)
+
+object H2OModelPayload {
+
+  def apply(inputStream: InputStream, name: String, version: String): H2OModelPayload = {
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/repository/H2OModelPayload.scala]: for me there is no sense in such construction - ou use case class or ordinal with its object and apply definition. Whar's the use case of this code? final case class is enough (hink if it should not be closeable like inptstream it contains???)

**Review by avan1235 — CHANGES_REQUESTED** [review]

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModel.scala (hunk 1 above)
Comment (kantoniak) [reply prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModel.scala:13]: `modelWrapper` is used in the signature provider.

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/repository/H2OModelPayload.scala (hunk 8 above)
Comment (kantoniak) [reply prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/repository/H2OModelPayload.scala]: I went with a construct similar to PMML repository. Changed to a minimal class for now.

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/repository/H2OModelPayload.scala [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package pl.touk.nussknacker.prinz.h2o.repository
+
+import java.io.InputStream
+
+// Tip: MOJO readers also support java.net.URL in place of InputStream
+class H2OModelPayload(var inputStream: InputStream, var name: String, var version: String)
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/repository/H2OModelPayload.scala]: case modifier was good here
 case class + object with apply function was for me without any sense

File: prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala [hunk 10]
```diff
@@ -1,3 +1,69 @@
...
+  private def getTransformer(modelCategory: ModelCategory): (AbstractPrediction => _) = modelCategory match {
+    case ModelCategory.AnomalyDetection => (p: AbstractPrediction) => p.asInstanceOf[AnomalyDetectionPrediction].isAnomaly
+    case ModelCategory.Binomial         => (p: AbstractPrediction) => p.asInstanceOf[BinomialModelPrediction].labelIndex
+    case ModelCategory.Multinomial      => (p: AbstractPrediction) => p.asInstanceOf[MultinomialModelPrediction].labelIndex
+    case ModelCategory.Ordinal          => (p: AbstractPrediction) => p.asInstanceOf[OrdinalModelPrediction].labelIndex
+    case ModelCategory.Clustering       => (p: AbstractPrediction) => p.asInstanceOf[ClusteringModelPrediction].cluster
+    case ModelCategory.KLime            => (p: AbstractPrediction) => p.asInstanceOf[KLimeModelPrediction].cluster
+    case ModelCategory.CoxPH            => (p: AbstractPrediction) => p.asInstanceOf[CoxPHModelPrediction].value
+    case ModelCategory.Regression       => (p: AbstractPrediction) => p.asInstanceOf[RegressionModelPrediction].value
+    case _ => throw new ModelRunException(s"ModelCategory ${modelCategory.toString()} not supported.")
```
Comment (avan1235) [inline prinz_h2o/src/main/scala/pl/touk/nussknacker/prinz/h2o/model/H2OModelInstance.scala]: .toString() is not needed here imo - clean your code

**Review by avan1235 — APPROVED** [review]
if you can please fix these small things and then merge

**Pull request merged**